File: 2022/01/30/c/balancedBinaryTree.c
// Program: https://leetcode.com/problems/balanced-binary-tree/
// Author: Andrew Jarombek
// Date: 1/30/2022

#include <stddef.h>
#include <stdbool.h>
#include <stdlib.h>

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

// Attempt #2 (Successful)

int dfs(struct TreeNode* root){
    if (root == NULL) {
        return 0;
    }

    int left = dfs(root->left);
    int right = dfs(root->right);

    if (left == -1 || right == -1 || abs(left - right) > 1) {
        return -1;
    }

    return 1 + (left > right ? left : right);
}

bool isBalanced(struct TreeNode* root){
    return dfs(root) != -1;
}

// Attempt #1

bool isBalanced1(struct TreeNode* root){
    if (root == NULL) {
        return true;
    }

    struct TreeNode* queue[5000];
    int start = 0;
    int end = 1;

    queue[0] = root;

    int highest = 0;
    int lowest = 0;
    int level = 1;

    while (start < end) {
        int newEnd = end;

        for (int i = start; i < end; i++) {
            struct TreeNode* node = queue[i];

            if (node->left == NULL && node->right == NULL) {
                if (!highest || level < highest) {
                    highest = level;
                }

                lowest = level;
            } else if (node->left == NULL || node->right == NULL) {
                if (!highest) {
                    highest = level + 1;
                }

                lowest = level + 1;
            }

            if (node->left != NULL) {
                queue[newEnd] = node->left;
                newEnd++;
            }

            if (node->right != NULL) {
                queue[newEnd] = node->right;
                newEnd++;
            }
        }

        level++;
        start = end;
        end = newEnd;
    }

    return (lowest - highest) <= 1;
}

File: 2022/03/12/balancedBinaryTree.swift
// https://leetcode.com/problems/balanced-binary-tree/
// Author: Andrew Jarombek
// Date: 3/12/2022

public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init() { self.val = 0; self.left = nil; self.right = nil; }
    public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
    public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
        self.val = val
        self.left = left
        self.right = right
    }
}

class Solution {
    func isBalanced(_ root: TreeNode?) -> Bool {
        if root == nil {
            return true
        }

        let result = dfs(root!)
        return result.1
    }

    func dfs(_ node: TreeNode) -> (Int, Bool) {
        if node.left == nil && node.right == nil {
            return (1, true)
        }

        var left = (0, true)
        var right = (0, true)

        if node.left != nil {
            left = dfs(node.left!)
        }

        if node.right != nil {
            right = dfs(node.right!)
        }

        return (max(left.0, right.0) + 1, left.1 && right.1 && abs(left.0 - right.0) < 2)
    }
}